packages-exp/analytics-exp/src/initialize-ids.ts

```diff
@@ -0,0 +1,137 @@
+import { getId } from '@firebase/installations-exp';
+import { fetchDynamicConfigWithRetry } from './get-config';
+import { logger } from './logger';
+import { FirebaseApp } from '@firebase/app-types-exp';
+import {
+  isIndexedDBAvailable,
+  validateIndexedDBOpenable
+} from '@firebase/util';
+import { ERROR_FACTORY, AnalyticsError } from './errors';
+
+async function validateIndexedDB(): Promise<boolean> {
+  if (!isIndexedDBAvailable()) {
+    logger.warn(
+      ERROR_FACTORY.create(AnalyticsError.INDEXEDDB_UNAVAILABLE, {
+        errorInfo: 'IndexedDB is not available in this environment.'
+      }).message
+    );
+    return false;
+  } else {
+    try {
+      await validateIndexedDBOpenable();
+    } catch (e) {
+      logger.warn(
+        ERROR_FACTORY.create(AnalyticsError.INDEXEDDB_UNAVAILABLE, {
+          errorInfo: e
+        }).message
+      );
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Initialize the analytics instance in gtag.js by calling config command with fid.
+ *
+ * NOTE: We combine analytics initialization and setting fid together because we want fid to be
+ * part of the `page_view` event that's sent during the initialization
+ * @param app Firebase app
+ * @param gtagCore The gtag function that's not wrapped.
+ * @param dynamicConfigPromisesList Array of all dynamic config promises.
+ * @param measurementIdToAppId Maps measurementID to appID.
+ * @param installations FirebaseInstallations instance.
+ *
+ * @returns Measurement ID.
+ */
+export async function initializeIds(
+  app: FirebaseApp,
+  dynamicConfigPromisesList: Array<
+    Promise<DynamicConfig | MinimalDynamicConfig>
+  >,
+  measurementIdToAppId: { [key: string]: string },
+  installations: FirebaseInstallations,
+  gtagCore: Gtag
+): Promise<string> {
+  const dynamicConfigPromise = fetchDynamicConfigWithRetry(app);
+  // Once fetched, map measurementIds to appId, for ease of lookup in wrapped gtag function.
+  dynamicConfigPromise
+    .then(config => {
+      measurementIdToAppId[config.measurementId] = config.appId;
+      if (
+        app.options.measurementId &&
+        config.measurementId !== app.options.measurementId
+      ) {
+        logger.warn(
+          `The measurement ID in the local Firebase config (${app.options.measurementId})` +
+            ` does not match the measurement ID fetched from the server (${config.measurementId}).` +
+            ` To ensure analytics events are always sent to the correct Analytics property,` +
+            ` update the` +
+            ` measurement ID field in the local config or remove it from the local config.`
+        );
+      }
+    })
+    .catch(e => logger.error(e));
+  // Add to list to track state of all dynamic config promises.
+  dynamicConfigPromisesList.push(dynamicConfigPromise);
+
+  const fidPromise: Promise<string | undefined> = validateIndexedDB().then(
+    envIsValid => {
+      if (envIsValid) {
+        return getId(installations);
```

Comment:
The `installations` here is an internal installations instance where you can do `installations.getId()`. We never import/use functions from other Firebase SDKs directly. We rely entirely on the component framework for inter-SDK communication.

Reply:
Ah, I was confused about how it works.